man/disnix-client.8
.TH "disnix-client" "8" "June 2009" "Disnix" "System administration tools"
.SH NAME
disnix\-client \- Provides access to the disnix-service through the D-Bus interface
.SH SYNOPSIS
.B disnix\-client
{operation}
[
.B options
]
derivations
.PP
.SH DESCRIPTION
The command \fBdisnix\-client\fR provides access to a \fBdisnix\-service\fR instance running
on the same machine by connecting to the D-Bus system or session bus. 
.PP
In most cases this tool is only needed for debugging purposes, since it only uses the D-Bus
protocol and cannot connect to a remote machine. A more useful client for use in production
environments is: \fBdisnix\-ssh\-client\fR.

.SH OPERATIONS
.TP
\fB\-\-import\fR
Imports a given closure into the Nix store of the target machine. Optionally, it transfers the closure from this machine
to the target machine.
.TP
\fB\-\-export\fR
Exports the closure of a given Nix store path of the target machine into a file, and optionally downloads it.
.TP
\fB\-\-print\-invalid\fR
Prints all the paths that are not valid in the Nix store of the target machine.
.TP
\fB\-r\fR, \fB\-\-realise\fR
Realises the given store derivation on the target machine.
.TP
\fB\-\-set\fR
Creates a Disnix profile only containing the given derivation on the target machine.
.TP
\fB\-q\fR, \fB\-\-query\-installed\fR
Queries all the installed services on the given target machine.
.TP
\fB\-\-query\-requisites\fR
Queries all the requisites (intra-dependencies) of the given services on the target machine.
.TP
\fB\-\-collect\-garbage\fR
Collects garbage on the given target machine.
.TP
\fB\-\-activate\fR
Activates the given service on the target machine.
.TP
\fB\-\-deactivate\fR
Deactivates the given service on the target machine.
.TP
\fB\-\-lock\fR
Acquires a lock on a Disnix profile of the target machine.
.TP
\fB\-\-unlock\fR
Release the lock on a Disnix profile of the target machine.

.SH OPTIONS
.TP
\fB\-t\fR, \fB\-\-target\fR
This option does nothing. It is only defined to match the interfaces of other Disnix clients.
.TP
\fB\-\-session\-bus\fR
Connects to the session bus instead of the system bus. This is useful for testing/debugging purposes.

.SH Import operation
.TP
\fB\-\-localfile\fR
Specifies that the given closure is stored locally, so that it will be transfered to the target machine first.
.TP
\fB\-\-remotefile\fR
Specifies that the given closure is stored remotely, so that it does not have to be transfered to the target machine.

.SH Export operation
.TP
\fB\-\-localfile\fR
Specifies that the exported closure should be stored locally. This option does not download the closure.
.TP
\fB\-\-remotefile\fR
Specifies that the exported closure should be stored remotely. This option downloads the closure from the target machine.

.SH Set operation/Query installed operation
.TP
\fB\-p\fR, \fB\-\-profile\fR
Specifies the name of the profile that should be used. Defaults to: default.

.SH Collect garbage operation
.TP
\fB\-d\fR, \fB\-\-delete\-old\fR
Specifies that also the older generations of the profiles have to be deleted.

.SH Activation/Deactivation operation
.TP
\fB\-\-type\fR
Specifies the activation module that should be used, such as \fBecho\fR or \fBprocess\fR.
.TP
\fB\-\-arguments\fR
Specifies the arguments passed to the activation module, which is a string with key=value pairs.

.SH Lock/Unlock operation
.TP
\fB\-p\fR, \fB\-\-profile\fR
Specifies the name of the profile that should be used. Defaults to: default.

.SH SEE ALSO
.BR disnix-ssh-client (8),
.BR disnix-service (8),
.BR disnix-env (8),
.SH AUTHOR
Sander van der Burg <s.vanderburg@tudelft.nl>


src/dbus-service/disnix-client.c
/*
 * Disnix - A distributed application layer for Nix
 * Copyright (C) 2008-2010  Sander van der Burg
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <stdio.h>
#include <stdlib.h>
#define _GNU_SOURCE
#include <getopt.h>
#include <glib.h>
#include <dbus/dbus-glib.h>
#include "disnix-client.h"
#include "disnix-marshal.h"

typedef enum
{
    OP_NONE,
    OP_IMPORT,
    OP_EXPORT,
    OP_PRINT_INVALID,
    OP_REALISE,
    OP_SET,
    OP_QUERY_INSTALLED,
    OP_QUERY_REQUISITES,
    OP_COLLECT_GARBAGE,
    OP_ACTIVATE,
    OP_DEACTIVATE,
    OP_LOCK,
    OP_UNLOCK,
}
Operation;

static void print_usage()
{
    /* Print the usage */
    printf("Usage:\n");
    printf("disnix-client [--session-bus] operation\n\n");
    
    printf("Operations:\n");
    printf("--import [--localfile|--remotefile] derivations\n");
    printf("--export [--localfile|--remotefile] derivations\n");
    printf("--print-invalid derivations\n");
    printf("{-r|--realise} derivations\n");
    printf("--set [{-p|--profile} name] derivation\n");
    printf("{-q|--query-installed} [{-p|--profile} name]\n");
    printf("--query-requisites derivations\n");
    printf("--collect-garbage [{-d|--delete-old}]\n");
    printf("--activate --type type --arguments arguments derivation\n");
    printf("--deactivate --type type --arguments arguments derivation\n");
    printf("--lock [{-p|--profile} profile]\n");
    printf("--unlock [{-p|--profile} profile]\n");
    printf("{-h|--help}\n");
}

/* Signal handlers */

static void disnix_finish_signal_handler(DBusGProxy *proxy, const gint pid, gpointer user_data)
{
    gint my_pid = *((gint*)user_data);

    g_printerr("Received finish signal from pid: %d\n", pid);

    /* Stop the main loop if our job is done */
    if(pid == my_pid)
	exit(0);    
}

static void disnix_success_signal_handler(DBusGProxy *proxy, const gint pid, gchar **derivation, gpointer user_data)
{
    unsigned int i;
    gint my_pid = *((gint*)user_data);
    
    g_printerr("Received success signal from pid: %d\n", pid);
    
    for(i = 0; i < g_strv_length(derivation); i++)
	g_print("%s", derivation[i]);
	
    /* Stop the main loop if our job is done */
    if(pid == my_pid)
	exit(0);
}

static void disnix_failure_signal_handler(DBusGProxy *proxy, const gint pid, gpointer user_data)
{
    gint my_pid = *((gint*)user_data);
    
    g_printerr("Received failure signal from pid: %d\n", pid);
    
    /* Stop the main loop if our job is done */
    if(pid == my_pid)
	exit(1);
}

/**
 * Runs the client
 */
static int run_disnix_client(Operation operation, gchar **derivation, int session_bus, char *profile, int delete_old, gchar **arguments, char *type)
{
    /* The GObject representing a D-Bus connection. */
    DBusGConnection *bus;
    
    /* Proxy object representing the D-Bus service object. */
    DBusGProxy *remote_object;

    /* GMainLoop object */
    GMainLoop *mainloop;
    
    /* Captures the results of D-Bus operations */
    GError *error = NULL;    
    gint reply;
    
    /* Other declarations */
    gint pid;
    
    /* If no operation is specified we should quit */
    if(operation == OP_NONE)
    {
	g_printerr("No operation is specified!\n");
	return 1;
    }
    
    /* Initialize the GType/GObject system. */
    g_type_init();

    /* Create main loop */
    mainloop = g_main_loop_new(NULL, FALSE);
    if(mainloop == NULL)
    {
	g_printerr("Cannot create main loop.\n");
	g_strfreev(derivation);
	g_strfreev(arguments);
	return 1;
    }

    /* Connect to the session/system bus */
    
    if(session_bus)
    {
	g_printerr("Connecting to the session bus.\n");
	bus = dbus_g_bus_get (DBUS_BUS_SESSION, &error);
    }
    else
    {
	g_printerr("Connecting to the system bus.\n");
	bus = dbus_g_bus_get (DBUS_BUS_SYSTEM, &error);
    }
    
    if(!bus)
    {
        g_printerr("Cannot connect to session/system bus! Reason: %s\n", error->message);
	g_strfreev(derivation);
	g_strfreev(arguments);
        return 1;
    }
    
    /* Create the proxy object that will be used to access the object */
    
    g_printerr("Creating a Glib proxy object.\n");
    remote_object = dbus_g_proxy_new_for_name (bus,
					       "org.nixos.disnix.Disnix", /* name */
					       "/org/nixos/disnix/Disnix", /* Object path */
					       "org.nixos.disnix.Disnix"); /* Interface */
    if(remote_object == NULL)
    {
        g_printerr("Cannot create the proxy object! Reason: %s\n", error->message);
	g_strfreev(derivation);
	g_strfreev(arguments);
        return 1;
    }

    /* Register marshaller for the return values of the success signal */
    dbus_g_object_register_marshaller(marshal_VOID__INT_BOXED, G_TYPE_NONE, G_TYPE_INT, G_TYPE_STRV, G_TYPE_INVALID);
    
    /* Register the signatures for the signal handlers */

    g_printerr("Add the argument signatures for the signal handler\n");
    dbus_g_proxy_add_signal(remote_object, "finish", G_TYPE_INT, G_TYPE_INVALID);
    dbus_g_proxy_add_signal(remote_object, "success", G_TYPE_INT, G_TYPE_STRV, G_TYPE_INVALID);
    dbus_g_proxy_add_signal(remote_object, "failure", G_TYPE_INT, G_TYPE_INVALID);

    g_printerr("Register D-Bus signal handlers\n");
    dbus_g_proxy_connect_signal(remote_object, "finish", G_CALLBACK(disnix_finish_signal_handler), &pid, NULL);
    dbus_g_proxy_connect_signal(remote_object, "success", G_CALLBACK(disnix_success_signal_handler), &pid, NULL);
    dbus_g_proxy_connect_signal(remote_object, "failure", G_CALLBACK(disnix_failure_signal_handler), &pid, NULL);

    /* Receive a PID for the job we want to execute */
    org_nixos_disnix_Disnix_get_job_id(remote_object, &pid, &error);
    g_printerr("Assigned PID: %d\n", pid);

    /* Execute operation */
    g_printerr("Executing operation.\n");
    
    switch(operation)
    {
	case OP_IMPORT:
	    org_nixos_disnix_Disnix_import(remote_object, pid, derivation[0], &error);
	    break;
	case OP_EXPORT:
	    org_nixos_disnix_Disnix_export(remote_object, pid, (const gchar**) derivation, &error);
	    break;
	case OP_PRINT_INVALID:
	    org_nixos_disnix_Disnix_print_invalid(remote_object, pid, (const gchar**) derivation, &error);
	    break;
	case OP_REALISE:
	    org_nixos_disnix_Disnix_realise(remote_object, pid, (const gchar**) derivation, &error);
	    break;
	case OP_SET:
	    org_nixos_disnix_Disnix_set(remote_object, pid, profile, derivation[0], &error);
	    break;
	case OP_QUERY_INSTALLED:
	    org_nixos_disnix_Disnix_query_installed(remote_object, pid, profile, &error);
	    break;
	case OP_QUERY_REQUISITES:
	    org_nixos_disnix_Disnix_query_requisites(remote_object, pid, (const gchar**) derivation, &error);
	    break;
	case OP_COLLECT_GARBAGE:
	    org_nixos_disnix_Disnix_collect_garbage(remote_object, pid, delete_old, &error);
	    break;
	case OP_ACTIVATE:
	    org_nixos_disnix_Disnix_activate(remote_object, pid, derivation[0], type, (const gchar**) arguments, &error);
	    break;
	case OP_DEACTIVATE:
	    org_nixos_disnix_Disnix_deactivate(remote_object, pid, derivation[0], type, (const gchar**) arguments, &error);
	    break;
	case OP_LOCK:
	    org_nixos_disnix_Disnix_lock(remote_object, pid, profile, &error);
	    break;
	case OP_UNLOCK:
	    org_nixos_disnix_Disnix_unlock(remote_object, pid, profile, &error);
	    break;
	case OP_NONE:
	    g_printerr("ERROR: No operation specified!\n");	    
	    print_usage();
	    g_strfreev(derivation);
	    g_strfreev(arguments);
	    return 1;
    }

    if(error != NULL) 
    {
        g_printerr("Error while executing the operation! Reason: %s\n", error->message);
	g_strfreev(derivation);
	g_strfreev(arguments);
	return 1;
    }
    
    /* Run loop and wait for signals */
    g_main_loop_run(mainloop);
    
    /* Operation is finished */
    g_strfreev(derivation);
    g_strfreev(arguments);
    
    return EXIT_FAILURE;
}

int main(int argc, char *argv[])
{
    /* Declarations */
    int c, option_index = 0;
    struct option long_options[] =
    {
	{"import", no_argument, 0, 'I'},
	{"export", no_argument, 0, 'E'},
	{"print-invalid", no_argument, 0, 'P'},
	{"realise", no_argument, 0, 'r'},
	{"set", no_argument, 0, 'S'},
	{"query-installed", no_argument, 0, 'q'},
	{"query-requisites", no_argument, 0, 'Q'},
	{"collect-garbage", no_argument, 0, 'C'},
	{"activate", no_argument, 0, 'A'},
	{"deactivate", no_argument, 0, 'D'},
	{"lock", no_argument, 0, 'L'},
	{"unlock", no_argument, 0, 'U'},
	{"help", no_argument, 0, 'h'},
	{"target", required_argument, 0, 't'},
	{"localfile", no_argument, 0, 'l'},
	{"remotefile", no_argument, 0, 'R'},
	{"profile", required_argument, 0, 'p'},
	{"delete-old", no_argument, 0, 'd'},
	{"type", required_argument, 0, 'T'},
	{"arguments", required_argument, 0, 'a'},
	{"session-bus", no_argument, 0, 'b'},
	{0, 0, 0, 0}
    };

    /* Option value declarations */
    Operation operation = OP_NONE;
    char *target, *profile = "default", *type;
    gchar **derivation = NULL, **arguments = NULL;
    unsigned int derivation_size = 0, arguments_size = 0;
    int localfile = FALSE, remotefile = TRUE;
    int delete_old = FALSE, session_bus = FALSE;
    
    /* Parse command-line options */
    while((c = getopt_long(argc, argv, "rqt:p:dh", long_options, &option_index)) != -1)
    {
	switch(c)
	{	    
	    case 'I':
		operation = OP_IMPORT;
		break;
	    case 'E':
		operation = OP_EXPORT;		
		break;
	    case 'P':
		operation = OP_PRINT_INVALID;		
		break;
	    case 'r':
		operation = OP_REALISE;
		break;
	    case 'S':
		operation = OP_SET;
		break;
	    case 'q':
		operation = OP_QUERY_INSTALLED;
		break;
	    case 'Q':
		operation = OP_QUERY_REQUISITES;
		break;
	    case 'C':
		operation = OP_COLLECT_GARBAGE;
		break;
	    case 'A':
		operation = OP_ACTIVATE;
		break;
	    case 'D':
		operation = OP_DEACTIVATE;
		break;
	    case 'L':
		operation = OP_LOCK;
		break;
	    case 'U':
		operation = OP_UNLOCK;
		break;
	    case 't':
		target = optarg;
		break;
	    case 'l':
		localfile = TRUE;
		remotefile = FALSE;
		break;
	    case 'R':
		localfile = FALSE;
		remotefile = TRUE;
		break;
	    case 'p':
		profile = optarg;
		break;
	    case 'd':
		delete_old = TRUE;
		break;
	    case 'T':
		type = optarg;
		break;
	    case 'a':
		arguments = (gchar**)g_realloc(arguments, (arguments_size + 1) * sizeof(gchar*));
		arguments[arguments_size] = g_strdup(optarg);
		arguments_size++;
		break;
	    case 'b':
		session_bus = TRUE;
		break;
	    case 'h':
		print_usage();
		return 0;
	}
    }
    
    /* Validate non-options */
    while(optind < argc)
    {
	derivation = g_realloc(derivation, (derivation_size + 1) * sizeof(gchar*));
	derivation[derivation_size] = g_strdup(argv[optind]);
	derivation_size++;	
	optind++;
    }
    
    derivation = g_realloc(derivation, (derivation_size + 1) * sizeof(gchar*));
    derivation[derivation_size] = NULL;
    
    /* Execute Disnix client */
    return run_disnix_client(operation, derivation, session_bus, profile, delete_old, arguments, type);
}
